File: bindings/dart/lib/src/models/chapter.dart
class Chapter {
  Chapter({
    required this.index,
    required this.title,
    required this.url,
    required this.updatedAt,
  });

  final int index;
  final String title;
  final String url;
  final DateTime? updatedAt;

  factory Chapter.parse(Map<String, dynamic> map) {
    final updatedAtString =
        map['updated_at']['Utc'] ?? map['updated_at']['Local'];

    // FIXME: handle local date times appropriately
    final updatedAt =
        updatedAtString != null ? DateTime.parse(updatedAtString) : null;

    return Chapter(
      index: map['index'],
      title: map['title'],
      url: map['url'],
      updatedAt: updatedAt,
    );
  }
}

File: crates/core/src/data/chapter.rs
use chrono::{DateTime, NaiveDateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Chapter {
    pub index: i32,
    pub title: String,
    pub url: String,
    pub updated_at: Option<TaggedDateTime>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "snake_case")]
pub enum TaggedDateTime {
    Utc(DateTime<Utc>),
    Local(NaiveDateTime),
}

#[derive(Serialize, Deserialize, Default, Debug)]
pub struct Content {
    pub data: String,
}

impl From<String> for Content {
    fn from(value: String) -> Self {
        Content { data: value }
    }
}
